contributors/design-proposals/api-machinery/udp-portforwarding.md

```diff
@@ -0,0 +1,109 @@
+# UDP Port Forwarding in Kubernetes
+
+Author: @shiywang
+
+## Abstract
+
+This describes an approach for providing support for:
+
+- port forwarding to containers with UDP protocol
+
+## Background && Motivation
+
+Fixes user request of https://github.com/kubernetes/kubernetes/issues/47862
+
+
+## Constraints and Assumptions
+
+- We already have an implemention with SPDY for TCP, the UDP implement basically reuse the SPDY framework, so it would be UDP through SPDY, the change here is remove the stream copy since UDP doesn't not has the same stream attribute as SPDY/TCP, so UDP will use datagram and `write` and `send` to communicate, socat will use unix domain socket as a middle layer to cross the boundaries between pod and Kubelet. and since TCP/SPDY doesn't have a message boundaries but UDP does, we also need to design an simple protocol for how to divide message read from SPDY back and forth.
+
+![design digram](udp1.png)
+
+
+## Use Cases
+
+- As a user of a Kubernetes cluster, I want to be able to connect to local ports on my computer and have them forwarded to ports in the container with UDP.
+
+- Basically like this: `kubectl port-forward pod-name --protocol=udp   8888:10001`
+
+## Process Flow
+
+### Port Forwarding Flow
+1. The Kubectl connects to the Kubernetes Master to initiate a remote command execution
+request.
+2. The Master proxies the request to the Kubelet where the container lives.
+3. The Kubectl listens on each specified local port, awaiting reading package.
+4. The udp client send package to one of the local listening ports in Kubectl, Kubectl parse the package.
+4. The Kubectl notifies the Kubelet of the new connection when the rmAddr is different. (similar to tcp per connections per data stream, since udp is connectionless we use rmAddr to distinguish)
+5. The Kubelet executes nsenter + socat and use unix domain socket listening and forwarding back and forth between the client and the port in the container.
+
+
+## Design Considerations
+
+### Backward Compatibility
+
+#### old kubectl - new api-server/kubelet
+ 
+Fully compatible with out compatibility design.
+
+#### new kubectl - old api-server/kubelet
+
+Add an `PortForwardProtocolV2Name` in client side, as well as deal it in api-server side, so the old server deosn't recognize will throw an error.
+
+### Customized Protocol
+
+The customized protocol is pretty simple here, it contains two parts: the content length and actual data.
+
+![simple protocol](udp2.png)
```

Comment:
I would do this as a markdown table instead of an image.

Reply:
It might be worth noting that traditional UDP datagrams have a maximum length of ~ 65535 bytes (i.e. a 2-byte unsigned integer), whereas the length field of a jumbogram may use up to 4 bytes (2^32 - 1).

Reply:
Yes, from practical perspective, package length more than MTU length will be dropped, and even we take ipv6 into consideration (16-bits payload length), so I think `uint32` is enough.